Add tutorial for caching of static assets (#4342)
umbraco/UmbracoDocs: Tutorials/Cache-Static-Assets/index.md

        {
            _globalSettings = globalSettings.Value;
            _hostingEnvironment = hostingEnvironment;
        }

        public void Configure(StaticFileOptions options)
            => options.OnPrepareResponse = ctx =>
            {
                string backOfficePath = _globalSettings.GetBackOfficePath(_hostingEnvironment);

                // Exclude Umbraco backoffice assets
                if (ctx.Context.Request.Path.StartsWithSegments(backOfficePath))
                {
                    return;
                }

                // Set headers for specific file extensions
                var fileExtension = Path.GetExtension(ctx.File.Name);
                if (_fileExtensions.Contains(fileExtension))
                {
                    var headers = ctx.Context.Response.GetTypedHeaders();

                    // Update or set Cache-Control header
                    var cacheControl = headers.CacheControl ?? new CacheControlHeaderValue();
                    cacheControl.Public = true;
                    cacheControl.MaxAge = TimeSpan.FromDays(365);
                    headers.CacheControl = cacheControl;
                }
            };
    }
}
```

> ## Cache images in ImageSharp
> For setting `Cache-Control` max-age header for images processed by the ImageSharp middleware, you can set the `Umbraco:CMS:Imaging:Cache:BrowserMaxAge` setting.
> https://our.umbraco.com/Documentation/Reference/Configuration/ImagingSettings/#browser-max-age

Reviewer: ```suggestion
## Cache images in ImageSharp

For setting `Cache-Control` max-age header for images processed by the ImageSharp middleware, you can set the `Umbraco:CMS:Imaging:Cache:BrowserMaxAge` setting.

See the [Images Settings](https://our.umbraco.com/Documentation/Reference/Configuration/ImagingSettings/#browser-max-age) article for more information.
```

---
versionFrom: 9.3.0
---

# Cache static assets

Example to cache static assets by file extension, but excluding Umbraco BackOffice assets.

```csharp
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;
using Microsoft.Net.Http.Headers;
using System;
using System.Collections.Generic;
using System.IO;
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.Configuration.Models;
using Umbraco.Cms.Core.DependencyInjection;

> using Umbraco.Cms.Core.Hosting;
> using Umbraco.Extensions;

Reviewer 2: For V10 there is an ambiguous reference between different `IHostingEnvironment` implementations; best make it explicit:

```suggestion
using Umbraco.Extensions;
using IHostingEnvironment = Umbraco.Cms.Core.Hosting.IHostingEnvironment;
```

Author: Seems to work fine to me without this change, which reference `Umbraco.Cms.Core.Hosting.Umbraco.Cms.Core.Hosting` , but yes if one include `Microsoft.AspNetCore.Hosting` or maybe in global usings is may be ambiguous.

> 

Reviewer 2: The code samples usually has a namespace declaration. While not mandatory it is good practice to add.

```suggestion

namespace My.Site.Composers;

```

public class StaticFilesComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
        => builder.Services.AddTransient<IConfigureOptions<StaticFileOptions>, ConfigureStaticFileOptions>();

    private class ConfigureStaticFileOptions : IConfigureOptions<StaticFileOptions>
    {

>         private readonly IHostingEnvironment _hostingEnvironment;
>         private readonly GlobalSettings _globalSettings;
>         private static readonly HashSet<string> _fileExtensions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
>         {
>             ".ico",
>             ".css",
>             ".js",
>             ".svg",
>             ".woff2"
>         };

Reviewer 2: Consider removing the member variables `_hostingEnvironment` and `_globalSettings` in favor of calculating the back-office path in the constructor.

Also consider renaming the `_fileExtensions` member for clarity and maybe add a comment explaining what it holds.

Lastly it is common practice to have static members at the top of the class.

```suggestion
        // These are the extensions of the file types we want to cache (add and remove as you see fit)
        private static readonly HashSet<string> _cachedFileExtensions = new(StringComparer.OrdinalIgnoreCase)
        {
            ".ico",
            ".css",
            ".js",
            ".svg",
            ".woff2"
        };

        private readonly string _backOfficePath;
```

>         public ConfigureStaticFileOptions(IOptions<GlobalSettings> globalSettings, IHostingEnvironment hostingEnvironment)
>         {
>             _globalSettings = globalSettings.Value;
>             _hostingEnvironment = hostingEnvironment;
>         }
> 

Reviewer 2: See comment above.

```suggestion
        public ConfigureStaticFileOptions(IOptions<GlobalSettings> globalSettings, IHostingEnvironment hostingEnvironment)
            => _backOfficePath = globalSettings.Value.GetBackOfficePath(hostingEnvironment);

```

        public void Configure(StaticFileOptions options)
            => options.OnPrepareResponse = ctx =>
            {

>                 string backOfficePath = _globalSettings.GetBackOfficePath(_hostingEnvironment);
> 

Reviewer 2: See comment above

```suggestion
```

                // Exclude Umbraco backoffice assets
                if (ctx.Context.Request.Path.StartsWithSegments(backOfficePath))
                {
                    return;
                }

                // Set headers for specific file extensions
                var fileExtension = Path.GetExtension(ctx.File.Name);

>                 if (_fileExtensions.Contains(fileExtension))

Reviewer 2: See comment above

```suggestion
                if (_cachedFileExtensions.Contains(fileExtension))
```

                {

>                     var headers = ctx.Context.Response.GetTypedHeaders();

Reviewer 2: Consider using explicit type here

```suggestion
                    ResponseHeaders headers = ctx.Context.Response.GetTypedHeaders();
```

Author: It seems `ResponseHeaders` it need to reference to `Microsoft.AspNetCore.Http.Headers` as well.

                    // Update or set Cache-Control header

>                     var cacheControl = headers.CacheControl ?? new CacheControlHeaderValue();

Reviewer 2: Consider using explicit type here

```suggestion
                    CacheControlHeaderValue cacheControl = headers.CacheControl ?? new CacheControlHeaderValue();
```

Author: @kjac in that case we could also just write:

```
CacheControlHeaderValue cacheControl = headers.CacheControl ?? new();
```

                    cacheControl.Public = true;
                    cacheControl.MaxAge = TimeSpan.FromDays(365);
                    headers.CacheControl = cacheControl;
                }
            };
    }
}
```

## Cache images in ImageSharp

For setting `Cache-Control` max-age header for images processed by the ImageSharp middleware, you can set the `Umbraco:CMS:Imaging:Cache:BrowserMaxAge` setting.

> See the [Images Settings](https://our.umbraco.com/Documentation/Reference/Configuration/ImagingSettings/#browser-max-age) article for more information.

Reviewer 2: The anchor link may result in people losing context of the linked article, consider removing it.

```suggestion
See the [Images Settings](https://our.umbraco.com/Documentation/Reference/Configuration/ImagingSettings/) article for more information.
```

> 
> ```json  
> "Umbraco": {
>     "CMS": {
>       "Imaging": {
>         "Cache": {
>           "BrowserMaxAge": "365.00:00:00",
>           "CacheMaxAge": "365.00:00:00",
>           "CachedNameLength": 8,
>           "CacheFolder": "~/umbraco/Data/TEMP/MediaCache"
>         },
>         "Resize": {
>           "MaxWidth": 5000,
>           "MaxHeight": 5000
>         }
>       }
>     }
>   }
> ```

Reviewer 2: This is duplicated from the [linked article](https://our.umbraco.com/Documentation/Reference/Configuration/ImagingSettings/) - consider removing it to avoid stale content.

```suggestion
```
